src/vivarium/framework/values.py

```diff
@@ -193,15 +229,66 @@ class Pipeline:
 
     """
 
-    def __init__(self):
-        self.name = None
-        self.source = None
-        self.mutators = []
-        self.combiner = None
-        self.post_processor = None
-        self.manager = None
+    def __init__(self) -> None:
+        self.name: str | None = None
+        self._source: ValueSource | None = None
+        self.mutators: list[ValueMutator] = []
+        self._combiner: ValueCombiner | None = None
+        self.post_processor: PostProcessor | None = None
+        self._manager: ValuesManager | None = None
+
+    def _get_attr_error(self, attribute: str) -> str:
+        return (
+            f"The pipeline for {self.name} has no {attribute}. This likely means "
+            f"you are attempting to modify a value that hasn't been created."
+        )
+
+    def _set_attr_error(self, attribute: str, new_value: Any) -> str:
+        current_value = getattr(self, f"_{attribute}")
+        return (
+            f"A second component is attempting to set the {attribute} for pipeline {self.name} "
+            f"with {new_value}, but it already has a {attribute}: {current_value}."
+        )
+
+    @property
+    def source(self) -> ValueSource:
+        if self._source is None:
+            raise DynamicValueError(self._get_attr_error("source"))
+        return self._source
```

Comment:
We could reduce the code duplication like this?

```python
T = TypeVar("T")

def _get_property(self, property: T | None, property_name: str) -> T:
    if property is None:
        raise DynamicValueError(self._get_attr_error(property_name))
    return property

def _set_property(self, property_name: str, new_value: Any) -> None:
    private_name = f"_{property_name}"
    old_value = getattr(self, private_name)
    if old_value is not None:
        raise DynamicValueError(self._set_attr_error(property_name, old_value, _new_value))
    setattr(self, private_name, new_value)

@property
def source(self) -> ValueSource:
    return self._get_property(self._source, "source")

@source.setter
def source(self, source: ValueSource) -> None:
    _set_property("source", source)
```

Reply:
Ah, i tried to go all the way with this but I was missing the TypeVar part, so it was throwing errors. thanks!